lib/ts/MemArena.cc

```diff
@@ -0,0 +1,179 @@
+/** @file
+
+    MemArena memory allocator. Chunks of memory are allocated, frozen into generations and
+     thawed away when unused.
+
+    @section license License
+
+    Licensed to the Apache Software Foundation (ASF) under one
+    or more contributor license agreements.  See the NOTICE file
+    distributed with this work for additional information
+    regarding copyright ownership.  The ASF licenses this file
+    to you under the Apache License, Version 2.0 (the
+    "License"); you may not use this file except in compliance
+    with the License.  You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+    Unless required by applicable law or agreed to in writing, software
+    distributed under the License is distributed on an "AS IS" BASIS,
+    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+    See the License for the specific language governing permissions and
+    limitations under the License.
+ */
+
+#include "MemArena.h"
+#include <ts/ink_memory.h>
+#include <ts/ink_assert.h>
+
+using namespace ts;
+
+inline void
+MemArena::MemArenaBlock::alloc(size_t n)
+{
+  ink_assert(n > 0);
+
+  data      = ats_malloc(n);
+  size      = n;
+  allocated = 0;
+  next      = nullptr;
+}
+
+inline void
+MemArena::MemArenaBlock::free()
+{
+  ats_free(data);
+  size      = 0;
+  allocated = 0;
+  data      = nullptr;
+  next      = nullptr;
+}
+
+/**
+    Marks a new generation. Caller should set next_block_size before creating a new generation.
+    Previous generation is implicitly frozen.
+ */
+inline void
+MemArena::create_new_generation()
+{
+  if (!next_block_size) {
+    next_block_size = arena_size * 2;
+  }
+
+  MemArenaBlock *next_gen = new MemArenaBlock();
+  next_gen->alloc(next_block_size);
+
+  // update generation marker
+  generation->next = next_gen;
+  generation       = next_gen;
+  arena_size       = next_block_size;
+}
+
+MemArena::MemArena()
+{
+  MemArena(arena_size); // Default size
+}
+
+MemArena::MemArena(size_t n)
+{
+  allocations = new MemArenaBlock();
+  allocations->alloc(n);
+  arena_size = n;
+  generation = allocations; // first generation
+}
+
+MemArena::~MemArena()
+{
+  MemArenaBlock *tmp = allocations;
+  MemArenaBlock *next;
+  while (tmp) {
+    next = tmp->next;
+    tmp->free();
+    delete tmp;
+
+    tmp = next;
+  }
+
+  allocations = nullptr;
+  generation  = nullptr;
+}
+
+MemSpan
+MemArena::alloc(size_t n)
+{
+  std::lock_guard<std::mutex> lock(arena_mutex);
+
+  // frozen. must allocate in a new generation.
+  if (next_generation) {
+    create_new_generation();
+    next_generation = false;
+  }
+
+  // keep creating generations memory blocks until we have something large enough
```

Comment:
What happens to all those intermediate blocks? Is it really efficient to allocate these until one is big enough?

Reply:
Default is to try double in size. If it's not big enough, instead of doubling in size, the next internal block will just be `n` bytes.